velox/functions/sparksql/String.cpp

```diff
@@ -103,6 +103,229 @@ class Length : public exec::VectorFunction {
+  // They are concatenated in advance.
+  constantStrings.reserve(numArgs - 1);
+  // For each array arg, save rawSizes, rawOffsets, indices, and elements
+  // BaseBector.
+  std::vector<const vector_size_t*> rawSizesVector;
+  std::vector<const vector_size_t*> rawOffsetsVector;
+  std::vector<const vector_size_t*> indicesVector;
+  std::vector<DecodedVector> decodedVectors;
+
+  for (auto i = 1; i < numArgs; ++i) {
+    if (args[i] && args[i]->typeKind() == TypeKind::ARRAY) {
+      exec::LocalDecodedVector arrayHolder(context, *args[i], rows);
+      auto& arrayDecoded = *arrayHolder.get();
+      auto baseArray = arrayDecoded.base()->as<ArrayVector>();
+      rawSizesVector.push_back(baseArray->rawSizes());
+      rawOffsetsVector.push_back(baseArray->rawOffsets());
+      indicesVector.push_back(arrayDecoded.indices());
+      auto elements = baseArray->elements();
+      exec::LocalSelectivityVector nestedRows(context, elements->size());
+      nestedRows.get()->setAll();
+      exec::LocalDecodedVector elementsHolder(
+          context, *elements, *nestedRows.get());
+      auto& elementsDecoded = *elementsHolder.get();
+      decodedVectors.push_back(std::move(elementsDecoded));
+      continue;
+    }
+    // Handles string arg.
+    argMapping.push_back(i);
+    if (!separator.has_value()) {
+      // Cannot concat consecutive constant string args in advance.
+      constantStrings.push_back("");
+      continue;
+    }
+    if (args[i] && args[i]->as<ConstantVector<StringView>>() &&
+        !args[i]->as<ConstantVector<StringView>>()->isNullAt(0)) {
+      std::ostringstream out;
+      out << args[i]->as<ConstantVector<StringView>>()->valueAt(0);
+      column_index_t j = i + 1;
+      // Concat constant string args in advance.
+      for (; j < numArgs; ++j) {
+        if (!args[j] || args[j]->typeKind() == TypeKind::ARRAY ||
+            !args[j]->as<ConstantVector<StringView>>() ||
+            args[j]->as<ConstantVector<StringView>>()->isNullAt(0)) {
+          break;
+        }
+        out << separator.value()
+            << args[j]->as<ConstantVector<StringView>>()->valueAt(0);
+      }
+      constantStrings.emplace_back(out.str());
+      i = j - 1;
+    } else {
+      constantStrings.push_back("");
+    }
+  }
+
+  // Number of string columns after combined consecutive constant ones.
+  auto numStringCols = constantStrings.size();
+  // For column string arg decoding.
+  std::vector<exec::LocalDecodedVector> decodedStringArgs;
+  decodedStringArgs.reserve(numStringCols);
+  for (auto i = 0; i < numStringCols; ++i) {
+    if (constantStrings[i].empty()) {
+      auto index = argMapping[i];
+      decodedStringArgs.emplace_back(context, *args[index], rows);
+    }
+  }
+  exec::LocalDecodedVector separatorDecoded(context);
+  if (!separator.has_value()) {
+    separatorDecoded = exec::LocalDecodedVector(context, *args[0], rows);
+  }
+
+  // Calculate the total number of bytes in the result.
```

Comment:
It seems several helper functions could be extracted to achieve a single purpose.